File: tags/sdk-1.0-before-interface-removal/implementation/UrakawaToolkit/media/data/IMediaDataFactory.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace urakawa.media.data
{
	/// <summary>
	/// Factory creating <see cref="IMediaData"/>s
	/// </summary>
	public interface IMediaDataFactory
	{
		/// <summary>
		/// Gets the <see cref="IMediaDataPresentation"/> asociated <c>this</c>.
		/// Convenience for <c>this.getMediaDataManager().getPresentation()</c>
		/// </summary>
		/// <returns>The <see cref="IMediaDataPresentation"/></returns>
		/// <exception cref="exception.IsNotInitializedException">
		/// Thrown when no <see cref="IMediaDataPresentation"/> has yet been associated with <c>this</c>
		/// </exception>
		IMediaDataPresentation getPresentation();

		/// <summary>
		/// Gets the <see cref="IMediaDataManager"/> associated with <c>this</c> 
		/// (via. the <see cref="IMediaDataPresentation"/> associated with <c>this</c>.
		/// This is convenience for <c>this.<see cref="getPresentation"/>().<see cref="getMediaDataManager"/>getMediaDataManager()</c>
		/// </summary>
		/// <returns>The <see cref="IMediaDataManager"/></returns>
		IMediaDataManager getMediaDataManager();

		/// <summary>
		/// Initializer associating the <c>this</c> with an owning media data manager
		/// </summary>
		/// <param name="mngr">The manager</param>
		void setMediaDataManager(IMediaDataManager mngr);

		/// <summary>
		/// Creates a <see cref="IMediaData"/>
		/// </summary>
		/// <param name="xukLocalName"></param>
		/// <param name="xukNamespaceUri"></param>
		/// <returns></returns>
		IMediaData createMediaData(string xukLocalName, string xukNamespaceUri);

		/// <summary>
		/// Creates a <see cref="IMediaData"/> a given type
		/// </summary>
		/// <param name="mediaType">The given type</param>
		/// <returns>The created media data</returns>
		IMediaData createMediaData(Type mediaType);

		
	}
}

File: java/core/src/org/daisy/urakawa/media/data/IMediaDataFactory.java
package org.daisy.urakawa.media.data;

import org.daisy.urakawa.IWithPresentation;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.media.data.audio.IAudioMediaData;
import org.daisy.urakawa.media.data.audio.codec.WavAudioMediaData;
import org.daisy.urakawa.xuk.IXukAble;

/**
 * <p>
 * This is the factory that creates
 * {@link org.daisy.urakawa.media.data.IMediaData} instances.
 * </p>
 * <p>
 * The returned object is managed by its associated manager.
 * </p>
 * 
 * @depend - Create - org.daisy.urakawa.media.data.IMediaData
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Aggregation 1 org.daisy.urakawa.IPresentation
 */
public interface IMediaDataFactory extends IXukAble, IWithPresentation {
	/**
	 * Convenience method to obtain the manager from the IPresentation instance.
	 * 
	 * @return manager
	 */
	public IMediaDataManager getMediaDataManager();

	/**
	 * <p>
	 * Creates a new media data, managed.
	 * </p>
	 * <p>
	 * This factory method takes arguments to specify the exact type of object
	 * to create, given by the unique QName (XML Qualified Name) used in the XUK
	 * serialization format. This method can be used to generate instances of
	 * subclasses of the base object type.
	 * </p>
	 * 
	 * @param xukLocalName
	 *            cannot be null, cannot be empty string.
	 * @param xukNamespaceURI
	 *            cannot be null, but can be empty string.
	 * @return can return null (in case the QName specification does not match
	 *         any supported type).
	 * @tagvalue Exceptions "MethodParameterIsNull-MethodParameterIsEmptyString"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 * @throws MethodParameterIsEmptyStringException
	 *             Empty string '' method parameter is forbidden:
	 *             <b>xukLocalName</b>
	 */
	IMediaData createMediaData(String xukLocalName, String xukNamespaceURI)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException;

	/**
	 * <p>
	 * Creates a new media data, managed.
	 * </p>
	 * <p>
	 * This factory method takes a single argument to specify the exact type of
	 * object to create.
	 * </p>
	 * 
	 * @param mediaType
	 * @return can return null (in case the given argument does not match any
	 *         supported type).
	 * @tagvalue Exceptions "MethodParameterIsNull"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 */
	IMediaData createMediaData(Class<IMediaData> mediaType)
			throws MethodParameterIsNullException;

	/**
	 * @return an instance of IAudioMediaData
	 */
	public IAudioMediaData createAudioMediaData();

	/**
	 * @return an instance of WavAudioMediaData
	 */
	public WavAudioMediaData createWavAudioMediaData();
}
